users/tests/test_api.py

```diff
@@ -0,0 +1,136 @@
+    put_data = {**PROFILE_TEST_DATA, "first_name": "Maija", "address": "Kauppakatu 23"}
+    response = APIClient().put(url, put_data)
+    assert response.status_code == 200
+    assert response.data == put_data
+    profile.refresh_from_db()
+    for attr in ["first_name", "last_name", "email"]:
+        assert str(getattr(profile.user, attr)) == str(put_data.pop(attr))
+    for attr, value in put_data.items():
+        assert str(getattr(profile, attr)) == str(value)
+
+
+@pytest.mark.django_db
+def test_profile_patch():
+    profile = _create_profile_from_data(PROFILE_TEST_DATA)
+    url = reverse("v1/profiles:profile-detail", args=(PROFILE_TEST_DATA["id"],))
+    patch_data = {"first_name": "Maija", "address": "Kauppakatu 23"}
+    response = APIClient().patch(url, patch_data)
+    assert response.status_code == 200
+    assert response.data == {**PROFILE_TEST_DATA, **patch_data}
+    profile.refresh_from_db()
+    assert profile.address == patch_data["address"]
+    assert profile.user.first_name == patch_data["first_name"]
+
+
+@pytest.mark.django_db
+def test_profile_delete():
+    profile = _create_profile_from_data(PROFILE_TEST_DATA)
+    url = reverse("v1/profiles:profile-detail", args=(PROFILE_TEST_DATA["id"],))
+    response = APIClient().delete(url)
+    assert response.status_code == 204
+    assert not User.objects.filter(pk=profile.user.pk).exists()
+    assert not Profile.objects.filter(pk=profile.pk).exists()
+
+
+@pytest.mark.django_db
+@patch.object(Profile.objects, "create")
+def test_profile_creation_is_rolled_back_on_profile_create_error(
+    patched_profile_objects_create,
+):
+    patched_profile_objects_create.side_effect = Exception
+    with pytest.raises(Exception):
+        APIClient().post(reverse("v1/profiles:profile-list"), PROFILE_TEST_DATA)
+    assert not User.objects.filter(
+        first_name=PROFILE_TEST_DATA["first_name"],
+        last_name=PROFILE_TEST_DATA["last_name"],
+        email=PROFILE_TEST_DATA["email"],
+    ).exists()
+    assert not Profile.objects.filter(pk=PROFILE_TEST_DATA["id"]).exists()
+
+
+@pytest.mark.django_db
+def test_profile_update_is_rolled_back_on_profile_save_error():
+    profile = _create_profile_from_data(PROFILE_TEST_DATA)
+    with patch.object(Profile, "save") as patched_profile_save:
+        patched_profile_save.side_effect = Exception
+        with pytest.raises(Exception):
+            url = reverse("v1/profiles:profile-detail", args=(PROFILE_TEST_DATA["id"],))
+            put_data = {
+                **PROFILE_TEST_DATA,
+                "first_name": "Maija",
+                "address": "Kauppakatu 23",
+            }
+            APIClient().put(url, put_data)
+    profile.refresh_from_db()
+    assert profile.user.first_name == PROFILE_TEST_DATA["first_name"]
+    assert profile.address == PROFILE_TEST_DATA["address"]
+
+
+def _create_profile_from_data(profile_data: dict) -> Profile:
```

Comment:
Should this be better as a fixture?

Reply:
Modified to a fixture. 👍 It actually made some of the tests a bit cleaner.